Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,892 @@
+newtype StateF s a = StateF { runStateF :: s -> (a, s) }
+  deriving stock Functor
+```
+
+(note that we're calling it `StateF` for "state functor")
+
+We can also define the typical functions for working with state:
+
+```haskell
+getF :: StateF s s
+getF = StateF $ \s -> (s, s)
+
+putF :: s -> StateF s ()
+putF s = StateF $ const ((), s)
+```
+
+Now we can define a "free state" monad:
+
+```haskell
+type State s = Free (StateF s)
+```
+
+We'll also need to lift our state operations into the monad.
+We could do it manually, of course.
+We have `StateF s a`, and we need to get `StateF s (Free (StateF s) a)`, and wrap it in `Free`.
+To lift any `a` into `Free f a` we just need to apply `Pure`:
+
+```haskell
+get :: State s s
+get = Free $ Pure <$> getF
+```
+
+But since it is a common pattern, the `free` library provides a function to do this for us:
+
+```haskell
+liftF :: (Functor f, MonadFree f m) => f a -> m a
+```
+
+`MonadFree` is just an MTL-style type class for different encodings of the free monad.
+In our case, we can pretend that
+
+```haskell
+liftF :: Functor f => f a -> Free f a
+-- The implementation of this simplified version
+-- is rather straightforward, as discussed above,
+-- so it is left as an exercise for the reader.
+```
+
+Hence, we can define
+
+```haskell
+get :: State s s
+get = liftF getF
+
+put :: s -> State s ()
+put = liftF . putF
+```
+
+and, like magic, we have a `State` monad:
+
+```haskell
+someComputation :: State Int ()
+someComputation = do
+  i <- get
+  put $ i + 1
+  pure ()
+```
+
+There is a catch, however.
+This code will happily compile, but it doesn't do anything interesting.
+Indeed, we didn't define the meaning of the computation, only its form, and we can't directly use `runStateF` to "run" our free state monad.
+
+This means we need to write an interpreter, so let's do just that.
+Remember that `Free` has two constructors, `Pure` and `Free`, so we need to pattern-match on those:
+
+```haskell
+runState :: State s a -> s -> (a, s)
+runState (Pure x) s = (x, s)
+runState (Free f) s =
+  let (m, s') = runStateF f s
+  in runState m s'
+```
+
+If you're familiar with the usual state monad, you may have noticed that we've essentially moved the implementation of `>>=` to `runState`.
+The free monad doesn't specify the meaning of monadic actions, so we have to decide what those actions mean when we're running it.
+
+To illustrate this point, we're going to write another interpreter, which is essentially a pretty-printer for the flow of a `State` computation:
+
+```haskell
+printState :: (Show s, Show a) => State s a -> s -> String
+printState (Pure x) s = "pure (" <> show x <> "," <> show s <> ")"
+printState (Free m) s =
+  let (x, s') = runStateF m s
+  in "state change " <> show s <> " -> " <> show s' <> "\n"
+    <> printState x s'
+```
+
+If we run `someComputation` defined above through this interpreter, for example by invoking `printState someComputation 1`, we'll get the following output:
```

Comment:
For example needs to be surrounded by commas. But I think it's easier to just take it out, don' t think it does anything here.